pkg/storage/http/http.go

```diff
@@ -0,0 +1,147 @@
+package http
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+)
+
+// ModuleStore represents an HTTP-backed storage backend.
+type ModuleStore struct {
+	url      string // base url
+	username string // http basic auth username
+	password string // http basic auth password
+	timeout  time.Duration
+	client   *http.Client
+}
+
+// New returns a new HTTP backed storage
+// that satisfies the Backend interface.
+func New(conf *config.HTTPConfig, timeout time.Duration) (*ModuleStore, error) {
+	const op errors.Op = "http.NewStorage"
+	if conf == nil {
+		return nil, errors.E(op, "No HTTP Configuration provided")
+	}
+	ms := &ModuleStore{url: conf.BaseURL, username: conf.Username, password: conf.Password, timeout: timeout, client: http.DefaultClient}
+
+	err := ms.connect()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return ms, nil
+}
+
+func (s *ModuleStore) connect() error {
+	const op errors.Op = "http.connect"
+
+	// I guess just GET the base URL and see if it 401's?
```

Comment:
Sounds good to me! Maybe we should require that it returns a 204. What do you think?

Reply:
Given the following facts:

- I use the base URL for this check
- The base URL will necessarily refer to a directory
- Elsewhere, `GET`s on directories are expected to yield a directory listing

...it seems a little inconsistent to hope for `No Content` here. For example, my backend of interest (Artifactory) responds with `200 OK` and a directory listing.

You do make a good point though: right now it fails on status codes `>= 400` but I could tighten that to `>= 200 && < 300` to more clearly reflect the expectation of this check. Thoughts?

Reply:
Got it. Then I have a more generic question - do you intend to build a generic HTTP storage driver in this PR, or one that can use Artifactory as a storage driver?

The two goals overlap, but I think there will need to be a spec for the driver (it looks like you started on that in here 😁) if you want to do the former. With the latter, I think it is fine to make everything in this PR work with Artifactory, without worrying about other possible backends.

Personally, I think it would be best to go with the latter (Artifactory-specific) since that achieves your immediate needs. Later on, we could of course always build a more generic HTTP backend with a spec and all...